### DISCUSSION TITLE
ValueError: 'id' mismatch in returned update response

### DISCUSSION THREAD

**Discussion by Tre-Seibert** [ecederstrand/exchangelib#1241] [Q&A] [answered]
In the code snippet below I am updating fields of a contact with the display name = "Test Contact". I get the following, 'ValueError: 'id' mismatch in returned update response' error. The error occurs when I contact.save(). However, the fields for the contact are still updated in Outlook despite the error. Any ideas how to bypass this error?

```
    # Retrieve each contact from TB Client to compare to report csv file.
    for contact in fTBClients.filter(display_name = "Test Contact"):
        contact.company_name = "Test Corp."
        contact.customerIDProperty = "TEST"
        contact.save()
        print("Contact saved")
```

```
Traceback (most recent call last):
  File "/Users/treseibert/Library/CloudStorage/OneDrive-SharedLibraries-TB/DEV - Documents/Office 365 Sync/app2.py", line 84, in <module>
    main()
  File "/Users/treseibert/Library/CloudStorage/OneDrive-SharedLibraries-TB/DEV - Documents/Office 365 Sync/app2.py", line 63, in main
    contact.save()
  File "/usr/local/lib/python3.11/site-packages/exchangelib/items/item.py", line 155, in save
    raise ValueError("'id' mismatch in returned update response")
ValueError: 'id' mismatch in returned update response
```

**Comment by ecederstrand** [comment]
In Exchange, all items have an ID and a ChangeKey value. When you update an item, the ID is supposed to stay the same, and the ChangeKey is supposed to change. On update, the server will return the ID and the ChangeKey of the newly updated item.

In this case, it looks like the server also returns a new ID for the updated item. You can confirm whether this is the case if you [enable debug logging](https://ecederstrand.github.io/exchangelib/#troubleshooting) and capture the request and response XML documents. If the server indeed returns a new ID on update, then we'll just have to relax this `ValueError` check.

**Reply by Tre-Seibert** [reply]
Thank you for this. Apologies for late closing.

**Comment by Tre-Seibert** [comment, marked as answer]
If anyone has this issue, the fix is simple. Locate items.py. In the save function, comment the following line:

`
#raise ValueError("'id' mismatch in returned update response")
`

Also, adjust the indentation of this line to be inside the if self.id conditional:

`
self._id = self.ID_ELEMENT_CLS(item_id, changekey)
`

Resulting in the following save function:


    def save(self, update_fields=None, conflict_resolution=AUTO_RESOLVE, send_meeting_invitations=SEND_TO_NONE):
        from .task import Task

        if self.id:
            item_id, changekey = self._update(
                update_fieldnames=update_fields,
                message_disposition=SAVE_ONLY,
                conflict_resolution=conflict_resolution,
                send_meeting_invitations=send_meeting_invitations,
            )
            if (
                self.id != item_id
                and not isinstance(self._id, (OccurrenceItemId, RecurringMasterItemId))
                and not isinstance(self, Task)
            ):
                # When we update an item with an OccurrenceItemId as ID, EWS returns the ID of the occurrence, so
                # the ID of this item changes.
                #
                # When we update certain fields on a task, the ID may change. A full description is available at
                # https://docs.microsoft.com/en-us/exchange/client-developer/web-service-reference/updateitem-operation-task
                #raise ValueError("'id' mismatch in returned update response")
                # Don't check that changekeys are different. No-op saves will sometimes leave the changekey intact
                self._id = self.ID_ELEMENT_CLS(item_id, changekey)
        else:
           ....unchanged

**Reply by ecederstrand** [reply]
If you can work out in which situation this happens, I'm happy to adjust the save() method to handle this.
